### PR TITLE
Split the persistence thread and the message handler.

### PR DESCRIPTION
**Pull request by baldersheim**
- Let FileStorManager own and control the Component and PersistenceHandler
  separately from the Persistence thread.
- Let FileStorManager allocate and control stripe assignment.

@vekterli and @geirst PR

### CODE REVIEW

File: storage/src/tests/persistence/filestorage/filestormanagertest.cpp [hunk 1]
```diff
@@ -721,12 +714,14 @@ TEST_F(FileStorManagerTest, priority) {
     FileStorHandlerImpl filestorHandler(messageSender, metrics, _node->getComponentRegister());
     ServiceLayerComponent component(_node->getComponentRegister(), "test");
     BucketOwnershipNotifier bucketOwnershipNotifier(component, messageSender);
-    std::unique_ptr<DiskThread> thread(createThread(
-            *config, *_node, _node->getPersistenceProvider(),
-            filestorHandler, bucketOwnershipNotifier, *metrics.disks[0]->threads[0]));
-    std::unique_ptr<DiskThread> thread2(createThread(
-            *config, *_node, _node->getPersistenceProvider(),
-            filestorHandler, bucketOwnershipNotifier, *metrics.disks[0]->threads[1]));
+    vespa::config::content::StorFilestorConfig cfg;
```
Comment (geirst) [inline storage/src/tests/persistence/filestorage/filestormanagertest.cpp:717]: The setup code per test function is very similar, and we could bundle it together in a helper class. Something to consider for upcoming work.

File: storage/src/vespa/storage/persistence/persistencehandler.cpp [hunk 2]
```diff
@@ -0,0 +1,162 @@
+// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
```
Comment (geirst) [inline storage/src/vespa/storage/persistence/persistencehandler.cpp:1]: Please update copyright.

File: storage/src/vespa/storage/persistence/persistencehandler.cpp [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+
+#include <vespa/storage/common/bucketoperationlogger.h>
+#include <vespa/document/fieldset/fieldsetrepo.h>
+#include <vespa/document/base/exceptions.h>
+#include <vespa/vespalib/util/exceptions.h>
+#include <vespa/vespalib/util/isequencedtaskexecutor.h>
+#include <thread>
+
+#include <vespa/log/log.h>
+LOG_SETUP(".persistence.thread");
```
Comment (geirst) [inline storage/src/vespa/storage/persistence/persistencehandler.cpp:13]: Please update log target.

File: storage/src/vespa/storage/persistence/persistencehandler.cpp [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+        case ReadBucketList::ID:
+            return _simpleHandler.handleReadBucketList(static_cast<ReadBucketList&>(msg), std::move(tracker));
+        case ReadBucketInfo::ID:
+            return _simpleHandler.handleReadBucketInfo(static_cast<ReadBucketInfo&>(msg), std::move(tracker));
+        case InternalBucketJoinCommand::ID:
+            return _splitJoinHandler.handleInternalBucketJoin(static_cast<InternalBucketJoinCommand&>(msg), std::move(tracker));
+        case RecheckBucketInfoCommand::ID:
+            return _splitJoinHandler.handleRecheckBucketInfo(static_cast<RecheckBucketInfoCommand&>(msg), std::move(tracker));
+        default:
+            LOG(warning, "Persistence thread received unhandled internal command %s", msg.toString().c_str());
```
Comment (geirst) [inline storage/src/vespa/storage/persistence/persistencehandler.cpp:92]: Please update message.

File: storage/src/vespa/storage/persistence/persistencehandler.h [hunk 5]
```diff
@@ -0,0 +1,50 @@
+// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
```
Comment (geirst) [inline storage/src/vespa/storage/persistence/persistencehandler.h:1]: Please update copyright.

File: storage/src/vespa/storage/persistence/persistencehandler.h [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+#include "simplemessagehandler.h"
+#include <vespa/storage/common/storagecomponent.h>
+#include <vespa/vespalib/util/isequencedtaskexecutor.h>
+#include <vespa/config-stor-filestor.h>
+
+namespace storage {
+
+class BucketOwnershipNotifier;
+
+class PersistenceHandler : public Types
```
Comment (geirst) [inline storage/src/vespa/storage/persistence/persistencehandler.h:20]: Consider adding class comment.

**Review by geirst — APPROVED** [review]
👍

File: storage/src/vespa/storage/persistence/persistencehandler.h [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+{
+public:
+    PersistenceHandler(vespalib::ISequencedTaskExecutor &, ServiceLayerComponent & component,
+                      const vespa::config::content::StorFilestorConfig &, spi::PersistenceProvider &,
```
Comment (vekterli) [inline storage/src/vespa/storage/persistence/persistencehandler.h:24]: Nit: indentation off by one

**Review by vekterli — APPROVED** [review]
:+1:

File: storage/src/tests/persistence/filestorage/filestormanagertest.cpp (hunk 1 above)
Comment (baldersheim) [reply storage/src/tests/persistence/filestorage/filestormanagertest.cpp:717]: I have noticed the same, will do on next encounter.

File: storage/src/vespa/storage/persistence/persistencehandler.cpp (hunk 2 above)
Comment (baldersheim) [reply storage/src/vespa/storage/persistence/persistencehandler.cpp:1]: Fixed

File: storage/src/vespa/storage/persistence/persistencehandler.cpp (hunk 3 above)
Comment (baldersheim) [reply storage/src/vespa/storage/persistence/persistencehandler.cpp:13]: Fixed

File: storage/src/vespa/storage/persistence/persistencehandler.cpp (hunk 4 above)
Comment (baldersheim) [reply storage/src/vespa/storage/persistence/persistencehandler.cpp:92]: Fixed

File: storage/src/vespa/storage/persistence/persistencehandler.h (hunk 5 above)
Comment (baldersheim) [reply storage/src/vespa/storage/persistence/persistencehandler.h:1]: Fixed

File: storage/src/vespa/storage/persistence/persistencehandler.h (hunk 6 above)
Comment (baldersheim) [reply storage/src/vespa/storage/persistence/persistencehandler.h:20]: Done

File: storage/src/vespa/storage/persistence/filestorage/filestormanager.cpp [hunk 8]
```diff
@@ -133,8 +140,14 @@ FileStorManager::configure(std::unique_ptr<vespa::config::content::StorFilestorC
...
-            _threads.push_back(std::make_shared<PersistenceThread>(*_sequencedExecutor, _compReg, *_config, *_provider,
-                                                                   *_filestorHandler, *_bucketOwnershipNotifier, *_metrics->disks[0]->threads[j]));
+            _persistenceComponents.push_back(std::make_unique<ServiceLayerComponent>(_compReg, createThreadName(j)));
+            _persistenceHandlers.push_back(
+                    std::make_unique<PersistenceHandler>(*_sequencedExecutor,
+                                                         *_persistenceComponents.back(),
+                                                         *_config, *_provider, *_filestorHandler,
+                                                         *_bucketOwnershipNotifier, *_metrics->disks[0]->threads[j]));
+            _threads.push_back(std::make_unique<PersistenceThread>(*_persistenceHandlers.back(), *_filestorHandler,
+                                                                   j % numStripes, _component));
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply storage/src/vespa/storage/persistence/filestorage/filestormanager.cpp:150]: Fixed

File: storage/src/vespa/storage/persistence/persistencehandler.cpp [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+#include <vespa/log/log.h>
+LOG_SETUP(".persistence.thread");
+
+using vespalib::make_string_short::fmt;
+using to_str = vespalib::string;
+
+namespace storage {
+
+PersistenceHandler::PersistenceHandler(vespalib::ISequencedTaskExecutor & sequencedExecutor,
+                                     ServiceLayerComponent & component,
```
[the remark this answers is not in the document]
Comment (baldersheim) [reply storage/src/vespa/storage/persistence/persistencehandler.cpp:21]: Fixed

File: storage/src/vespa/storage/persistence/persistencehandler.h (hunk 7 above)
Comment (baldersheim) [reply storage/src/vespa/storage/persistence/persistencehandler.h:24]: Fixed

**Pull request merged**